### ISSUE TITLE
Build issues with Xcode 11.4 as dependency of firefox-ios

### ISSUE THREAD

**Issue by armcknight** [mozilla/application-services#2872]
I'm trying to run `bootstrap.sh --force` in the Firefox iOS project (https://github.com/mozilla-mobile/firefox-ios/blob/master/Cartfile#L19), which depends on this as a Carthage dependency. The build is failing:
```
Build Failed
	Task failed with exit code 65:
	/usr/bin/xcrun xcodebuild -project /[redacted]/firefox-ios/Carthage/Checkouts/application-services/megazords/ios/MozillaAppServices.xcodeproj -scheme MozillaAppServices -configuration Release -derivedDataPath /[redacted]/Library/Caches/org.carthage.CarthageKit/DerivedData/11.4_11E146/application-services/v0.54.0 -sdk iphoneos ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES archive -archivePath /var/folders/fj/pd0rb2nj1xj05rhjmjxkv4rc0000gn/T/application-services SKIP_INSTALL=YES GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=NO CLANG_ENABLE_CODE_COVERAGE=NO STRIP_INSTALLED_PRODUCT=NO (launched in /[redacted]/firefox-ios/Carthage/Checkouts/application-services)

This usually indicates that project itself failed to compile. Please check the xcodebuild log for more details: /var/folders/fj/pd0rb2nj1xj05rhjmjxkv4rc0000gn/T/carthage-xcodebuild.y6idMe.log
```

There seem to be several underlying problems: 
1. the `gyp` dependency is not satisfied by the time `carthage bootstrap` is run
2. an assumption about the location of `cargo`
3. running `cargo` fails to compile, or to resolve dependencies, I can't tell

# 1. `gyp` dependency

I was able to pretty easily follow the directions to install `gyp` on my machine, but it would be nice if this could be automated as some sort of bootstrap.sh script like the Firefox project has: https://github.com/mozilla-mobile/firefox-ios/blob/master/bootstrap.sh. It took me a while to figure out this was causing the initial failure to update Firefox iOS' Carthage dependencies. 

Maybe this could even be part of their bootstrap script? Happy to file an issue against that repo.

# 2. `cargo` location

Once I resolved the `gyp` issue, I got a new error:
```
~/.cargo/bin/cargo: No such file or directory
```
occurring here: https://github.com/mozilla/application-services/blob/master/build-scripts/xc-universal-binary.sh#L61

I don't think this script should make assumptions like this about where things are outside of the repo root. I changed the line from
```
"${HOME}"/.cargo/bin/cargo build --locked -p "${FFI_TARGET}" --lib ${RELFLAG} --target "${IOS_TRIPLES[${i}]}"
```
to
```
cargo build --locked -p "${FFI_TARGET}" --lib ${RELFLAG} --target "${IOS_TRIPLES[${i}]}"
```
and it worked. 

Happy to submit a PR if this change makes sense for the project.

# 3. Running `cargo`

After fixing the path to cargo above, the build script fails with the following messages:
```
# We can't use cargo lipo because we can't link to universal libraries :(
# https://github.com/rust-lang/rust/issues/55235
LIBS_ARCHS=("x86_64" "arm64")
+ LIBS_ARCHS=("x86_64" "arm64")
IOS_TRIPLES=("x86_64-apple-ios" "aarch64-apple-ios")
+ IOS_TRIPLES=("x86_64-apple-ios" "aarch64-apple-ios")
for i in "${!LIBS_ARCHS[@]}"; do
    LIB_ARCH=${LIBS_ARCHS[${i}]}
    env -i \
        PATH="${PATH}" \
        NSS_STATIC=1 \
        NSS_DIR="${LIBSDIR}/ios/${LIB_ARCH}/nss" \
        SQLCIPHER_LIB_DIR="${LIBSDIR}/ios/${LIB_ARCH}/sqlcipher/lib" \
        SQLCIPHER_INCLUDE_DIR="${LIBSDIR}/ios/${LIB_ARCH}/sqlcipher/include" \
        RUSTC_WRAPPER="${RUSTC_WRAPPER:-}" \
        SCCACHE_IDLE_TIMEOUT="${SCCACHE_IDLE_TIMEOUT:-}" \
        SCCACHE_CACHE_SIZE="${SCCACHE_CACHE_SIZE:-}" \
        SCCACHE_ERROR_LOG="${SCCACHE_ERROR_LOG:-}" \
        RUST_LOG="${RUST_LOG:-}" \
    cargo build --locked -p "${FFI_TARGET}" --lib ${RELFLAG} --target "${IOS_TRIPLES[${i}]}"
done
+ for i in '"${!LIBS_ARCHS[@]}"'
+ LIB_ARCH=x86_64
+ env -i PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/libexec:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/local/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/usr/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/usr/local/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/usr/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/usr/local/bin:/Applications/Xcode11_4/Xcode.app/Contents/Developer/Tools:/usr/local/opt/python@3.8/bin:/[redacted]/google-cloud-sdk/bin:/[redacted]/.swiftenv/shims:/[redacted]/.jenv/shims NSS_STATIC=1 NSS_DIR=/[redacted]/firefox-ios/Carthage/Checkouts/application-services/megazords/ios/../../libs/ios/x86_64/nss SQLCIPHER_LIB_DIR=/[redacted]/firefox-ios/Carthage/Checkouts/application-services/megazords/ios/../../libs/ios/x86_64/sqlcipher/lib SQLCIPHER_INCLUDE_DIR=/User[redacted]/firefox-ios/Carthage/Checkouts/application-services/megazords/ios/../../libs/ios/x86_64/sqlcipher/include RUSTC_WRAPPER= SCCACHE_IDLE_TIMEOUT= SCCACHE_CACHE_SIZE= SCCACHE_ERROR_LOG= RUST_LOG= cargo build --locked -p megazord_ios --lib --release --target x86_64-apple-ios
   Compiling libc v0.2.67
   Compiling proc-macro2 v1.0.9
   Compiling unicode-xid v0.2.0
   Compiling syn v1.0.16
   Compiling log v0.4.8
   Compiling cfg-if v0.1.10
   Compiling memchr v2.3.3
   Compiling getrandom v0.1.14
error[E0463]: can't find crate for `core`
  |
  = note: the `x86_64-apple-ios` target may not be installed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: could not compile `cfg-if`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

I'm not familiar with the rust ecosystem so I'm not sure how to dig deeper into this error.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1011)
┆Sprint: Backlog

**Comment by vladikoff** [comment]
hm I wonder if the commit in https://github.com/mozilla/application-services/commit/45c8c6f5446b083c81d925a7ade6a526b528a26f fixes this.

**Comment by eoger** [comment]
Hi @armcknight, thanks so much for the detailed ticket.

> Happy to submit a PR if this change makes sense for the project.

Yes I agree it makes sense, I wonder if we use this path because the `PATH` variable is not set properly on CI? Anyway it's fixable.

> error[E0463]: can't find crate for `core`

Have you tried running `rustup target add aarch64-apple-ios x86_64-apple-ios`?

Running `./libs/verify-ios-environment.sh` will ensure your environment is ready to build for iOS.

**Comment by eoger** [comment]
I'm also wondering if we should run the aforementioned script in the XCode build process of application services, so you can get a clear error of what's going on when building a-s for Firefox iOS from source because binaries are not available.

**Comment by eoger** [comment]
> hm I wonder if the commit in [45c8c6f](https://github.com/mozilla/application-services/commit/45c8c6f5446b083c81d925a7ade6a526b528a26f) fixes this.

If we made a release, it might allow the reporter to not have to build from source. However I'd like to wait till CircleCI releases their final xcode 11.4.0 image before doing so. (they are currently using Beta 3, which might still trigger a build from source if they are using the release version)

**Comment by armcknight** [comment]
Thanks for the quick responses! Good to know I can maybe wait for a new binary release instead of building from source.

I tried a few things, let me try to summarize:

1. Ran verify-ios-environment.sh, which stated `rustup` couldn't be found. 

2. Tried `brew install rustup` but that said it's already installed. 

3. Tried `rustup-init`:
```
❯ rustup-init
error: it looks like you have an existing installation of Rust at:
error: /usr/local/bin
error: rustup should not be installed alongside Rust. Please uninstall your existing Rust first.
error: Otherwise you may have confusion unless you are careful with your PATH
error: If you are sure that you want both rustup and your already installed Rust
error: then please restart the installation and pass `-y' to bypass this check.
error: cannot install while Rust is installed
```

4. Uninstalled the brew version of rust: `brew uninstall rust`

5. `rustup-init` now works 🎉 but it output this message: `To get started you need Cargo's bin directory ($HOME/.cargo/bin) in your PATH`. I tried entering a new login shell but still saw this:

```
❯ ./libs/verify-ios-environment.sh
Error: The Rust compiler needs to be installed. See https://rustup.rs/ for install instructions.
```

And verified that ~/.cargo/bin isn't on `$PATH`. FWIW I'm using fish shell.

Maybe there's a path issue, because it does exist at `~/.cargo/bin/rustc`, but it appears verify-ios-environment.sh doesn't prepend `${HOME}/.cargo/bin` onto `rustc` like `xc-universal-binary.sh` does for `cargo`.

6. I reverted my change to add back `${HOME}/.cargo/bin` in `xc-universal-binary.sh` and am rerunning `carthage build`, but in the end I still get this error:
```
error[E0463]: can't find crate for `core`
```

**Comment by eoger** [comment]
- I would suggest to install Rust using https://rustup.rs/ instead of through brew which isn't really supported I think.
- You **do** want `~/.cargo/bin` in your path for the rust stuff to work, I'm not sure how to do that using fish tho.
- `can't find crate for core` would be solved by running `./libs/verify-ios-environment.sh` successfully (it adds the proper rust targets to your rust installation).

**Comment by thomcc** [comment]
> You do want ~/.cargo/bin in your path for the rust stuff to work, I'm not sure how to do that using fish tho.

Something like `set -Ux fish_user_paths $fish_user_paths "$HOME/.cargo/bin"` would do it. You could also add it to paths but it's better to use fish_user_paths as a uvar.

**Comment by vladikoff** [comment]
Keep in mind our scripts might not test in Fish shell properly

**Comment by thomcc** [comment]
They don't... I've been meaning to write fish versions of them but... yeah.

**Comment by eoger** [comment]
We don't `source` anymore in the verify scripts, but I've been meaning to convert them to a nice python script asking you what you're trying to build, similar to what `./mach bootstrap` does for mozilla-central.

**Comment by armcknight** [comment]
I added `~/.cargo/bin` to `PATH` (not so different for fish shell, you just modify `~/.config/fish/fish.config` instead of e.g. `~/.bash_profile`, and some slight syntactical differences; or the way @thomcc posted) and ran verify-ios-environment.sh again:
```
❯ ./libs/verify-ios-environment.sh
info: downloading component 'rust-std' for 'aarch64-apple-ios'
info: installing component 'rust-std' for 'aarch64-apple-ios'
info: downloading component 'rust-std' for 'x86_64-apple-ios'
info: installing component 'rust-std' for 'x86_64-apple-ios'
Error: swift-protobuf needs to be installed. See https://github.com/apple/swift-protobuf#alternatively-install-via-homebrew for install instructions.
```

ran `brew install swift-protobuf`, then verify-ios-environment.sh asked me to install `xcpretty`. I use `rbenv` by default, so didn't feel great about installing it against bare system ruby, but doing that did satisfy verify-ios-environment.sh. (Personally, I would prefer if this project used a Gemfile to declare the xcpretty dependency, happy to send you a PR for this too.)

I also ran `brew uninstall rustup` and used the curl to bash from their website, as you suggested.

Running `carthage build` now succeeds 🎉
